File: gfx/sdl2/image.c
/*
 * Copyright (C) 2017 Manh Tran
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
#include <cherry/graphic/image.h>

#if IMAGE_TYPE == IMAGE_SDL2

#include <cherry/memory.h>
#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>

/*
 * allocate new image object from file path
 */
struct image *image_alloc(char *file)
{
        /* try create surface from file */
        SDL_Surface *tempimage = IMG_Load(file);
        if( ! tempimage) return 0;

        /* fix byte order */
        SDL_Surface *tempsurface;
        Uint32 rmask, gmask, bmask, amask;

        /* fix ios color */
#if SDL_BYTEORDER == SDL_BIG_ENDIAN
        rmask = 0xff000000;
        gmask = 0x00ff0000;
        bmask = 0x0000ff00;
        amask = 0x000000ff;
#else
        rmask = 0x000000ff;
        gmask = 0x0000ff00;
        bmask = 0x00ff0000;
        amask = 0xff000000;
#endif
        tempsurface = SDL_CreateRGBSurface(SDL_SWSURFACE, tempimage->w, tempimage->h, 32, rmask, gmask, bmask, amask);
        SDL_BlitSurface(tempimage, NULL, tempsurface, NULL);
        SDL_FreeSurface(tempimage);

        tempimage = tempsurface;

        struct image *p = smalloc(sizeof(struct image));
        p->ptr = tempimage;

        if(tempimage->format->BytesPerPixel == 3)
        {
                p->type = GL_RGB;
        }
        else
        {
                p->type = GL_RGBA;
        }

        return p;
}

/*
 * deallocate image object
 */
void image_free(struct image *p)
{
        if(p->ptr) {
                SDL_FreeSurface(p->ptr);
        }
        sfree(p);
}

/*
 * get image width
 */
u16 image_width(struct image *p)
{
        return ((SDL_Surface *)p->ptr)->w;
}

/*
 * get image height
 */
u16 image_height(struct image *p)
{
        return ((SDL_Surface *)p->ptr)->h;
}

/*
 * get image pixels
 */
void *image_pixels(struct image *p)
{
        return ((SDL_Surface *)p->ptr)->pixels;
}

/*
 * get image type
 */
i32 image_type(struct image *p)
{
        return p->type;
}

#endif

File: gfx/mtl/image.m
/*
 * Copyright (C) 2017 Manh Tran
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
#import <cherry/graphic/image.h>

#if OS == MTL && IMAGE_TYPE == IMAGE_OBJC
#import <cherry/graphic/metal.h>
#import <cherry/memory.h>
#import <cherry/stdio.h>
#import <cherry/string.h>
#import <UIKit/UIKit.h>

static void dataImage(struct image *p, UIImage *uiimage)
{
    CGImageRef imageRef = [uiimage CGImage];

    // Create a suitable bitmap context for extracting the bits of the image
    const NSUInteger width = CGImageGetWidth(imageRef);
    const NSUInteger height = CGImageGetHeight(imageRef);
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    uint8_t *rawData = (uint8_t *)malloc(height * width * 4);
    const NSUInteger bytesPerPixel = 4;
    const NSUInteger bytesPerRow = bytesPerPixel * width;
    const NSUInteger bitsPerComponent = 8;
    CGContextRef context = CGBitmapContextCreate(rawData, width, height,
                                                 bitsPerComponent, bytesPerRow, colorSpace,
                                                 kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
    CGColorSpaceRelease(colorSpace);

    CGContextTranslateCTM(context, 0, height);
    CGContextScaleCTM(context, 1, -1);

    CGContextDrawImage(context, CGRectMake(0, 0, width, height), imageRef);
    CGContextRelease(context);

    p->ptr = rawData;
    p->bytes_per_row = bytesPerRow;
}

/*
 * allocate new image object from file path
 */
struct image *image_alloc(char *file)
{
        /* create objc image */
        struct string *full_path = file_full_path(file);
        UIImage *uiimage = [UIImage imageWithContentsOfFile:[NSString stringWithUTF8String:full_path->ptr]];
        string_free(full_path);

        /* create application image */
        struct image *ret = smalloc(sizeof(struct image));
        ret->width = uiimage.size.width * uiimage.scale;
        ret->height = uiimage.size.height * uiimage.scale;
        dataImage(ret, uiimage);
        ret->type = MTLPixelFormatRGBA8Unorm;
        return ret;
}

/*
 * deallocate image object
 */
void image_free(struct image *p)
{
        free(p->ptr);
        sfree(p);
}

/*
 * get image width
 */
u16 image_width(struct image *p)
{
        return p->width;
}

/*
 * get image height
 */
u16 image_height(struct image *p)
{
        return p->height;
}

/*
 * get image pixels
 */
void *image_pixels(struct image *p)
{
        return p->ptr;
}

/*
 * get image type
 */
i32 image_type(struct image *p)
{
        return p->type;
}

#endif
